Description: do not verify unlocking passphrase in luksAddKey command
Author: Milan Broz <mbroz@redhat.com>
Last-Update: 2010-02-25

--- a/ChangeLog
+++ b/ChangeLog
@@ -1,3 +1,6 @@
+2010-02-25  Milan Broz  <mbroz@redhat.com>
+	* Do not verify unlocking passphrase in luksAddKey command.
+
 2010-01-17  Milan Broz  <mbroz@redhat.com>
 	* If gcrypt compiled with capabilities, document workaround for cryptsetup (see lib/gcrypt.c).
 	* Version 1.1.0.
--- a/lib/setup.c
+++ b/lib/setup.c
@@ -1511,7 +1511,7 @@
 				      keyfile, keyfile_size);
 		else
 			key_from_terminal(cd, _("Enter any passphrase: "),
-					&password, &passwordLen, 1);
+					&password, &passwordLen, 0);
 
 		if (!password)
 			return -EINVAL;
