From: Guilhem Moulin <guilhem@debian.org>
Date: Fri, 06 Jul 2018 16:32:10 +0200
Subject: Add configure option '--disable-internal-tests' to disable the internal test suite

---
 Makefile.am  |    5 ++++-
 configure.ac |    5 +++++
 2 files changed, 9 insertions(+), 1 deletion(-)

--- a/Makefile.am
+++ b/Makefile.am
@@ -1,5 +1,8 @@
 EXTRA_DIST = COPYING.LGPL FAQ docs misc
-SUBDIRS = po tests
+SUBDIRS = po
+if !SKIP_INTERNAL_TESTS
+SUBDIRS += tests
+endif
 TESTS =
 CLEANFILES =
 DISTCLEAN_TARGETS =
--- a/configure.ac
+++ b/configure.ac
@@ -567,6 +567,11 @@ test "x$with_tmpfilesdir" = "xno" || {
 }
 AM_CONDITIONAL(CRYPTSETUP_TMPFILE, test -n "$DEFAULT_TMPFILESDIR")
 
+AC_ARG_ENABLE([internal-tests],
+	AS_HELP_STRING([--disable-internal-tests],
+	[skip the internal test suite]),[], [skip_internal_tests=no])
+AM_CONDITIONAL(SKIP_INTERNAL_TESTS, test x$skip_internal_tests != xno)
+
 CS_STR_WITH([luks2-lock-path], [path to directory for LUKSv2 locks], [/run/cryptsetup])
 test -z "$with_luks2_lock_path" && with_luks2_lock_path=/run/cryptsetup
 CS_ABSPATH([${with_luks2_lock_path}],[with-luks2-lock-path])
